File: backend/internal/interface/http/entities/mapper/evaluation.go
package mapper

import (
	"github.com/green-ecolution/green-ecolution/backend/internal/domain/evaluation"
	"github.com/green-ecolution/green-ecolution/backend/internal/interface/http/entities"
	"github.com/green-ecolution/green-ecolution/backend/internal/utils"
)

func EvaluationFromResponse(source *evaluation.Evaluation) *entities.EvaluationResponse {
	if source == nil {
		return nil
	}
	return &entities.EvaluationResponse{
		TreeCount:             source.TreeCount,
		TreeClusterCount:      source.TreeClusterCount,
		SensorCount:           source.SensorCount,
		WateringPlanCount:     source.WateringPlanCount,
		TotalWaterConsumption: source.TotalWaterConsumption,
		UserWateringPlanCount: source.UserWateringPlanCount,
		VehicleEvaluation:     utils.MapSlice(source.VehicleEvaluation, vehicleEvaluationToResponse),
		RegionEvaluation:      utils.MapSlice(source.RegionEvaluation, regionEvaluationToResponse),
	}
}

func vehicleEvaluationToResponse(source *evaluation.VehicleEvaluation) *entities.VehicleEvaluationResponse {
	if source == nil {
		return nil
	}
	return &entities.VehicleEvaluationResponse{
		NumberPlate:       source.NumberPlate,
		WateringPlanCount: source.WateringPlanCount,
	}
}

func regionEvaluationToResponse(source *evaluation.RegionEvaluation) *entities.RegionEvaluationResponse {
	if source == nil {
		return nil
	}
	return &entities.RegionEvaluationResponse{
		Name:              source.Name,
		WateringPlanCount: source.WateringPlanCount,
	}
}

File: backend-rs/src/http/v1/dto/evaluation.rs
use serde::Serialize;

use crate::domain::evaluation::{RegionEvaluation, VehicleEvaluation};

#[derive(Debug, Serialize)]
pub struct RegionEvaluationResponse {
    pub name: String,
    pub watering_plan_count: u32,
}

impl From<&RegionEvaluation> for RegionEvaluationResponse {
    fn from(value: &RegionEvaluation) -> Self {
        Self {
            name: value.name.clone(),
            watering_plan_count: value.watering_plan_count,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct VehicleEvaluationResponse {
    pub number_plate: String,
    pub watering_plan_count: u32,
}

impl From<&VehicleEvaluation> for VehicleEvaluationResponse {
    fn from(value: &VehicleEvaluation) -> Self {
        Self {
            number_plate: value.number_plate.clone(),
            watering_plan_count: value.watering_plan_count,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct EvaluationResponse {
    pub tree_count: u32,
    pub treecluster_count: u32,
    pub sensor_count: u32,
    pub watering_plan_count: u32,
    pub user_watering_plan_count: u32,
    pub total_water_consumption: u64,
    pub region_evaluation: Vec<RegionEvaluationResponse>,
    pub vehicle_evaluation: Vec<VehicleEvaluationResponse>,
}
